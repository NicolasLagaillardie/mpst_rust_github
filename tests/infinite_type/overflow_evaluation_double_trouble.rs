use crossbeam_channel::{Receiver, Sender};
use std::marker;

// Trait
pub trait Session: marker::Sized + marker::Send {
    type Dual: Session<Dual = Self>;
}

// Struct
#[must_use]
#[derive(Debug)]
pub struct Send<T, S>
where
    T: marker::Send,
    S: Session,
    S::Dual: Session,
{
    pub channel: Sender<(T, S::Dual)>,
}

#[must_use]
#[derive(Debug)]
pub struct Recv<T, S>
where
    T: marker::Send,
    S: Session,
{
    pub channel: Receiver<(T, S)>,
}

#[must_use]
#[derive(Debug)]
pub struct End {
    pub sender: Sender<()>,
    pub receiver: Receiver<()>,
}

// impl
impl<T: marker::Send, S: Session> Session for Send<T, S> {
    type Dual = Recv<T, S::Dual>;
}

impl<T: marker::Send, S: Session> Session for Recv<T, S> {
    type Dual = Send<T, S::Dual>;
}

impl Session for End {
    type Dual = End;
}

// enum
enum Test0 {
    End(End),
    Branch(Recv<Test0, End>),
}
enum Test1 {
    End(End),
    Branch(Recv<i32, Recv<Test1, End>>),
}
enum Test2 {
    End(End),
    Branch(Send<i32, Recv<Test2, End>>),
}
enum Test3 {
    End(End),
    Branch(Recv<i32, Send<i32, Recv<Test3, End>>>),
}
enum Test4 {
    End(End),
    Branch(Send<i32, Send<i32, Recv<Test4, End>>>),
}
enum Test5 {
    End(End),
    Branch(Recv<i32, Recv<i32, Recv<Test5, End>>>),
}
enum Test6 {
    End(End),
    Branch(Send<i32, Recv<i32, Recv<Test6, End>>>),
}

fn main() {}
