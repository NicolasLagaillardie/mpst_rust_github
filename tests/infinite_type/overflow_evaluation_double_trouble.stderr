error[E0275]: overflow evaluating the requirement `Send<i32, Send<i32, Recv<Test4, End>>>: std::marker::Send`
  --> tests/infinite_type/overflow_evaluation_double_trouble.rs:70:12
   |
70 |     Branch(Send<i32, Send<i32, Recv<Test4, End>>>),
   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: required because it appears within the type `Test4`
  --> tests/infinite_type/overflow_evaluation_double_trouble.rs:68:6
   |
68 | enum Test4 {
   |      ^^^^^
note: required for `Recv<Test4, End>` to implement `Session`
  --> tests/infinite_type/overflow_evaluation_double_trouble.rs:43:35
   |
43 | impl<T: marker::Send, S: Session> Session for Recv<T, S> {
   |         ------------              ^^^^^^^     ^^^^^^^^^^
   |         |
   |         unsatisfied trait bound introduced here

error[E0275]: overflow evaluating the requirement `Send<i32, Recv<i32, Recv<Test6, End>>>: std::marker::Send`
  --> tests/infinite_type/overflow_evaluation_double_trouble.rs:78:12
   |
78 |     Branch(Send<i32, Recv<i32, Recv<Test6, End>>>),
   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: required because it appears within the type `Test6`
  --> tests/infinite_type/overflow_evaluation_double_trouble.rs:76:6
   |
76 | enum Test6 {
   |      ^^^^^
note: required for `Recv<Test6, End>` to implement `Session`
  --> tests/infinite_type/overflow_evaluation_double_trouble.rs:43:35
   |
43 | impl<T: marker::Send, S: Session> Session for Recv<T, S> {
   |         ------------              ^^^^^^^     ^^^^^^^^^^
   |         |
   |         unsatisfied trait bound introduced here
